iroh-net/src/magic_endpoint.rs

```diff
@@ -971,4 +981,96 @@ mod tests {
+            .insecure_skip_relay_cert_verify(true)
+            .alpns(vec![TEST_ALPN.to_vec()])
+            .relay_mode(RelayMode::Custom(relay_map))
+            .bind(0)
+            .await
+            .unwrap();
+
+        async fn handle_direct_conn(ep: MagicEndpoint, node_id: PublicKey) -> Result<()> {
+            let node_addr = NodeAddr::new(node_id);
+            ep.add_node_addr(node_addr)?;
+            let stream = ep.conn_type_stream(&node_id)?;
+            async fn get_direct_event(mut stream: ConnectionTypeStream) -> Result<()> {
+                while let Some(conn_type) = stream.next().await {
+                    tracing::debug!("got {conn_type:?}");
+                    if matches!(conn_type, ConnectionType::Direct(_)) {
+                        return Ok(());
+                    }
+                }
+                anyhow::bail!("conn_type stream ended before `ConnectionType::Direct`");
+            }
+            tokio::time::timeout(Duration::from_secs(15), get_direct_event(stream)).await??;
+            Ok(())
+        }
+
+        let ep1_nodeid = ep1.node_id();
+        let ep2_nodeid = ep2.node_id();
+
+        let ep1_nodeaddr = ep1.my_addr().await.unwrap();
+        tracing::info!(
+            "node id 1 {ep1_nodeid}, relay URL {:?}",
+            ep1_nodeaddr.relay_url()
+        );
+        tracing::info!("node id 2 {ep2_nodeid}");
+
+        let res_ep1 = tokio::spawn(handle_direct_conn(ep1.clone(), ep2_nodeid));
+
+        let ep1_abort_handle = res_ep1.abort_handle();
+        let _ep1_guard = CallOnDrop::new(move || {
+            ep1_abort_handle.abort();
+        });
+
+        let res_ep2 = tokio::spawn(handle_direct_conn(ep2.clone(), ep1_nodeid));
+        let ep2_abort_handle = res_ep2.abort_handle();
+        let _ep2_guard = CallOnDrop::new(move || {
+            ep2_abort_handle.abort();
+        });
+        async fn accept(ep: MagicEndpoint) -> (PublicKey, String, quinn::Connection) {
+            let incoming = ep.accept().await.unwrap();
+            accept_conn(incoming).await.unwrap()
+        }
+
+        // create a node addr with no direct connections
+        let ep1_nodeaddr = NodeAddr::from_parts(ep1_nodeid, Some(relay_url), vec![]);
+
+        let accept_res = tokio::spawn(accept(ep1.clone()));
+        let accept_abort_handle = accept_res.abort_handle();
+        let _accept_guard = CallOnDrop::new(move || {
+            accept_abort_handle.abort();
+        });
+
+        let _conn_2 = ep2.connect(ep1_nodeaddr, TEST_ALPN).await.unwrap();
+
+        let (got_id, _, _conn) = accept_res.await.unwrap();
+        assert_eq!(ep2_nodeid, got_id);
+
+        res_ep1.await.unwrap().unwrap();
+        res_ep2.await.unwrap().unwrap();
+
+        let conn_info = ep1.connection_info(ep2.node_id()).unwrap();
+        match conn_info.conn_type {
+            ConnectionType::Direct(_) => {}
+            _ => panic!("expected direct connection, got {:?}", conn_info.conn_type),
+        }
```

Comment:
I'm mildly worried that due to #2169 this could make this test flaky.  The `res_ep1` resulting in an `Ok` is already sufficient as assert maybe?

Reply:
oooo good call. I don't think we need this assertion anymore, I was using it when I was relying on the oneshot senders and wanted extra assurance that everything was sent at the right time.